src/handlers/install_handler.rs

```diff
@@ -275,43 +276,95 @@ async fn handle_building_from_source(
-                        "Git has to be installed in order to build neovim from source"
-                    ))
-                }
-                _ => return Err(anyhow!("Failed to clone neovim's repository")),
-            },
-        };
-        (child, false)
+
+    if let Err(error) = Command::new("git").output().await {
+        if error.kind() == std::io::ErrorKind::NotFound {
+            return Err(anyhow!(
+                "Git has to be installed in order to build neovim from source"
+            ));
+        }
+    }
+
+    // create neovim-git if it does not exist
+    let dirname = "neovim-git";
+    if let Err(error) = fs::metadata(dirname).await {
+        match error.kind() {
+            std::io::ErrorKind::NotFound => {
+                fs::create_dir(dirname).await?;
+            }
+            _ => return Err(anyhow!("unknown error: {}", error)),
+        }
+    }
+
+    env::set_current_dir(dirname)?; // cd into neovim-git
+
+    // check if repo is initialized
+    if let Err(error) = fs::metadata(".git").await {
+        match error.kind() {
+            std::io::ErrorKind::NotFound => {
+                Command::new("git")
+                    .arg("init")
+                    .stdout(Stdio::null())
+                    .spawn()?
+                    .wait()
+                    .await?;
+            }
+
+            _ => return Err(anyhow!("unknown error: {}", error)),
+        }
+    };
+
+    // check if repo has a remote
+    let remote = Command::new("git")
+        .arg("remote")
+        .arg("get-url")
+        .arg("origin")
+        .stdout(Stdio::null())
+        .spawn()?
+        .wait()
+        .await?;
+
+    if !remote.success() {
+        // add neovim's remote
+        Command::new("git")
+            .arg("remote")
+            .arg("add")
+            .arg("origin")
+            .arg("https://github.com/neovim/neovim.git")
+            .spawn()?
+            .wait()
+            .await?;
     } else {
-        env::set_current_dir("neovim-git")?; // cd into neovim-git
-        let child = match Command::new("git").arg("pull").spawn() {
-            Ok(value) => value,
-            Err(_) => return Err(anyhow!("Failed to pull upstream updates")),
-        };
-        (child, true)
+        // set neovim's remote otherwise
+        Command::new("git")
+            .arg("remote")
+            .arg("set-url")
+            .arg("origin")
+            .arg("https://github.com/neovim/neovim.git")
+            .spawn()?
+            .wait()
+            .await?;
     };
-    child.wait().await?;
-    if !is_installed {
-        env::set_current_dir("neovim-git")?; // cd into neovim-git
+    // fetch version from origin
+    let fetch_successful = Command::new("git")
+        .arg("fetch")
+        .arg("--depth")
+        .arg("1")
+        .arg("origin")
+        .arg(&version.non_parsed_string)
+        .spawn()?
+        .wait()
+        .await?.success();
+
+    if !fetch_successful {
```

Comment:
> do a check in the build function that if it's not the full hash abort the operation and tell them to provide the full hash.

I'm not sure what you mean by that. I've already added that. The error message could be improved by telling users to provide the full hash, sure.

Reply:
As in before the actual command of fetching the commit by hash. The success check would be just for good measure.

Reply:
fair enough. maybe add something like: "Try providing the full path" or something to the error message.

Reply:
Done.